### ISSUE TITLE
Centralize case sensitivity across the backends

### ISSUE THREAD

**Issue by dgutov** [company-mode/company-mode#364]
Maybe simply default to `completion-ignore-case`.

However, `ac-ignore-case` has an interesting value `smart`, which uses case sensitivity (or not) depending on whether the input contains a capital letter. That might warrant a new user option.

**Comment by stardiviner** [comment]
+1

**Comment by fommil** [comment]
:+1: Actually, for now I'd be happy enough knowing what the magic is to get `company-ispell` to respect the case of my `prefix`. e.g. if I type `Dys` I'd expect the completion to be `Dysfunction` and not `dysfunction`. I hope that makes sense.

**Comment by fommil** [comment]
if I'm reading this right https://github.com/company-mode/company-mode/blob/master/company.el#L1036 it looks like its not possible to preserve the case.

**Comment by dgutov** [comment]
Quite the opposite: since its `ignore-case` value is `keep-prefix`, the current prefix is substituted in at the beginning of each completion. And that's what I see in practice.

Maybe you're actually hitting a different backend? Would you care to post a screencast?

**Comment by fommil** [comment]
@dgutov you're right it works. I think I'm noticing that `complete-mode` is overriding the 

e.g. if I type `MONKEYlicious` in a buffer, `dabbrev-expand` will not do anything for `monkey` but will for `MONKEY`. However, company mode will offer `monkeylicious` as an expansion for `monkey`.

**Comment by dgutov** [comment]
@fommil You could try `(setq company-dabbrev-code-ignore-case 'keep-prefix)`.

This backend's implementation doesn't actually use the `dabbrev` package, so there's no overriding going on.

**Comment by fommil** [comment]
ok, `company-dabbrev-ignore-case nil` seems to help with false positives, but It's still wanting to insert a lowercased version for the remainder of the completion.

**Comment by dgutov** [comment]
What about not using nil, then, but the value I mentioned in the previous message?

**Comment by fommil** [comment]
that doesn't work either... the value I'm completing is all caps and when company completes on it the trailing part is lowercased.

**Comment by dgutov** [comment]
@fommil That is probably because of `company-dabbrev-downcase`. You should really check out the available user options first. In any case, please file a separate issue if you want to continue this discussion.

You may also want to look into the reason why `company-dabbrev` is being used, instead of `company-dabbrev-code`, which has default settings that must be more to your liking.

**Comment by fommil** [comment]
ok, thanks! That fixed it. I hadn't fully appreciated that you're not using `dabbrev` and I was confused why my default `dabbrev` settings weren't being respected. I can delete all my comments here to clear up the ticket.

**Comment by dgutov** [comment]
`dabbrev` is not written in a reusable-enough way for this, unfortunately.

I'll delete the offtopic comments later myself, don't worry.

**Comment by srustamo** [comment]
Upvoting 'smart' casing. Autocomplete offers both 'foobar' and 'Foobar' when entering 'foo'. Having this in company-mode would be excellent.

**Comment by stardiviner** [comment]
Any updates on this?
I wish it can works like this:
There are candidates:
```
.getcode
.getcache
.getAnnotation
.getapt
.GET_IDE
```

When I type `.geta` it can show candidates ignore case because I don't want to filter with case sensitive so early, until I really type `.getA` to trigger case sensitive explicitly.

**Comment by dgutov** [comment]
Let's discuss this.

Do we just introduce a variable and demand that all backends honor it? Especially the external ones.

I worry that some of them (e.g. ones using an external process) might have a problem implementing one or the other behavior. And if "don't ignore case" can be implemented via the other behavior, the reverse isn't true. Not sure what the guidelines for such backends should be.

Is this a real problem? Or are ycmd, jedi, ensime and LSP (the protocol) all able to handle both matching types?

**Comment by Sarcasm** [comment]
Regarding LSP, in the protocol I could not find any reference to case-sensitivity in the completion request section, but I guess this logic could be coded into the wrapper library.

**Comment by ptrv** [comment]
`ycmd` already works internally with "smart-case" sensitive filtering, so there is nothing to do on `company-ycmd` side for filtering, because the results from the server depend on the input, i.e. "foo" returns "Foo" and "foo", "Foo" returns "Foo" and "FOO" but not "foo".

**Comment by dgutov** [comment]
What if company wants it to always do case-insensitive matching? E.g., that's the user preference.

**Comment by ptrv** [comment]
> What if company wants it to always do case-insensitive matching?

Since the matching is not configurable in `ycmd` always case-insensitive matching would not work in `ycmd`.
